File: src/review/checks.ts
/**
 * docs/ 전체 품질 검토 체크 함수 모음.
 *
 * 8개 체크 함수가 JSONL emit으로 결과를 출력한다.
 */

import fs from "node:fs";
import path from "node:path";
import { error, verbose, warn } from "../core/emit.js";
import { parseFrontmatter } from "../core/frontmatter.js";

const REQUIRED_FM_ALL = new Set([
  "doc-id", "title", "description", "type", "status", "updated", "source",
]);
const REQUIRED_FM_API = new Set([
  "method", "path", "base-url", "category", "tags",
]);

function* iterMd(docs: string): Generator<string> {
  const entries = fs.readdirSync(docs, { withFileTypes: true });
  for (const e of entries) {
    const full = path.join(docs, e.name);
    if (e.isDirectory()) yield* iterMd(full);
    else if (e.isFile() && e.name.endsWith(".md")) yield full;
  }
}

function relTo(filePath: string, docs: string): string {
  try {
    return path.relative(docs, filePath).replace(/\\/g, "/");
  } catch {
    return filePath;
  }
}

// ── 1. 파일 수 / 구조 ────────────────────────────────────────────────────────

export function checkStructure(docs: string): void {
  const subdirs: Record<string, string> = {
    api: path.join(docs, "api"),
    schema: path.join(docs, "schema"),
    category: path.join(docs, "category"),
    guide: path.join(docs, "guide"),
  };

  let total = 0;
  const counts: Record<string, number> = {};
  for (const [name, d] of Object.entries(subdirs)) {
    const files = fs.existsSync(d)
      ? [...iterMd(d)]
      : [];
    counts[name] = files.length;
    total += files.length;
  }

  verbose("structure", { total, ...counts });

  for (const fname of ["llms.txt", "llms-full.txt"]) {
    const p = path.join(docs, fname);
    if (fs.existsSync(p)) {
      verbose("llms_file", { file: fname, size: fs.statSync(p).size });
    } else {
      warn("llms_file", { file: fname, msg: "파일 없음", ok: false });
    }
  }
}

// ── 2. 내부 링크 유효성 ───────────────────────────────────────────────────────

export function checkLinks(docs: string): number {
  let dead = 0;
  let checked = 0;
  const linkPat = /\[([^\]]*)\]\(([^)#\s][^)\s]*)\)/g;

  for (const md of iterMd(docs)) {
    const content = fs.readFileSync(md, "utf-8");
    const [, body] = parseFrontmatter(content);
    for (const m of body.matchAll(linkPat)) {
      const href = m[2];
      if (href.startsWith("http") || href.startsWith("mailto") || href.startsWith("#")) continue;
      const target = path.resolve(path.dirname(md), href);
      checked++;
      if (!fs.existsSync(target)) {
        warn("dead_link", { file: relTo(md, docs), href, ok: false });
        dead++;
      }
    }
  }

  verbose("links_done", { checked, dead, ok: dead === 0 });
  return dead;
}

// ── 3. frontmatter 완전성 ─────────────────────────────────────────────────────

export function checkFrontmatter(docs: string): void {
  let missingRequired = 0;
  let missingApi = 0;

  for (const md of iterMd(docs)) {
    const content = fs.readFileSync(md, "utf-8");
    const [fm] = parseFrontmatter(content);
    const rel = relTo(md, docs);

    if (!fm || Object.keys(fm).length === 0) {
      warn("frontmatter", { file: rel, msg: "frontmatter 없음", ok: false });
      missingRequired++;
      continue;
    }

    const pageType = String(fm["type"] ?? "");
    const required = new Set(REQUIRED_FM_ALL);
    if (pageType === "guide") required.delete("source");

    const missing = [...required].filter((k) => !(k in fm));
    if (missing.length > 0) {
      warn("frontmatter", {
        file: rel,
        msg: `필수 필드 누락: ${JSON.stringify(missing.sort())}`,
        ok: false,
      });
      missingRequired++;
    }

    if (pageType === "api-endpoint") {
      const m2 = [...REQUIRED_FM_API].filter((k) => !(k in fm));
      if (m2.length > 0) {
        warn("frontmatter_api", {
          file: rel,
          msg: `API 전용 필드 누락: ${JSON.stringify(m2.sort())}`,
          ok: false,
        });
        missingApi++;
      }
    }
  }

  verbose("frontmatter_done", {
    missing_required: missingRequired,
    missing_api: missingApi,
    ok: missingRequired === 0 && missingApi === 0,
  });
}

// ── 4. 내용 이상 패턴 ────────────────────────────────────────────────────────

export function checkContent(docs: string): void {
  const htmlPat = /<(?:div|span|table|tr|td|th|ul|li|p|br)\b/i;
  let tooShort = 0;
  let noH1 = 0;
  let multiH1 = 0;
  let hasHtml = 0;

  for (const md of iterMd(docs)) {
    const content = fs.readFileSync(md, "utf-8");
    const [, body] = parseFrontmatter(content);
    const rel = relTo(md, docs);

    if (body.trim().length < 100) {
      warn("content", { file: rel, msg: "본문 100자 미만", ok: false });
      tooShort++;
    }

    const h1s = [...body.matchAll(/^# .+/gm)];
    if (h1s.length === 0) {
      warn("content", { file: rel, msg: "H1 없음", ok: false });
      noH1++;
    } else if (h1s.length > 1) {
      warn("content", { file: rel, msg: `H1 중복: ${h1s.length}개`, ok: false });
      multiH1++;
    }

    if (htmlPat.test(body)) {
      warn("content", { file: rel, msg: "HTML 태그 잔존", ok: false });
      hasHtml++;
    }
  }

  verbose("content_done", {
    too_short: tooShort,
    no_h1: noH1,
    multi_h1: multiH1,
    has_html: hasHtml,
    ok: tooShort === 0 && noH1 === 0 && multiH1 === 0 && hasHtml === 0,
  });
}

// ── 5. llms.txt 정합성 ───────────────────────────────────────────────────────

export function checkLlms(docs: string): void {
  const llmsPath = path.join(docs, "llms.txt");
  if (!fs.existsSync(llmsPath)) {
    error("llms", { file: "llms.txt", msg: "파일 없음", ok: false });
    return;
  }

  const content = fs.readFileSync(llmsPath, "utf-8");
  const links = [...content.matchAll(/\(([^)]+\.md)\)/g)].map((m) => m[1]);
  const broken: string[] = [];
  for (const link of links) {
    const target = path.resolve(docs, link);
    if (!fs.existsSync(target)) broken.push(link);
  }

  const sections = [...content.matchAll(/^## .+/gm)];
  const entries = [...content.matchAll(/^- \[/gm)];

  for (const b of broken) {
    warn("llms_link", { file: "llms.txt", href: b, ok: false });
  }

  verbose("llms_done", {
    sections: sections.length,
    entries: entries.length,
    links: links.length,
    broken: broken.length,
    ok: broken.length === 0,
  });
}

// ── 6. 카테고리 인덱스 링크 ──────────────────────────────────────────────────

export function checkCategoryLinks(docs: string): void {
  const catDir = path.join(docs, "category");
  if (!fs.existsSync(catDir)) {
    warn("category_links", { msg: "category/ 없음", ok: false });
    return;
  }

  let total = 0;
  let dead = 0;
  const linkPat = /\[([^\]]+)\]\(([^)]+)\)/g;

  for (const entry of fs.readdirSync(catDir, { withFileTypes: true })) {
    if (!entry.isFile() || !entry.name.endsWith(".md")) continue;
    const md = path.join(catDir, entry.name);
    const content = fs.readFileSync(md, "utf-8");
    const [, body] = parseFrontmatter(content);
    const rel = relTo(md, docs);

    for (const m of body.matchAll(linkPat)) {
      const href = m[2];
      if (href.startsWith("http") || href.startsWith("#")) continue;
      total++;
      const target = path.resolve(path.dirname(md), href);
      if (!fs.existsSync(target)) {
        warn("category_link", { file: rel, href, ok: false });
        dead++;
      }
    }
  }

  verbose("category_links_done", { total, dead, ok: dead === 0 });
}

// ── 7. doc-id 중복 ───────────────────────────────────────────────────────────

export function checkDocIdUniqueness(docs: string): void {
  const idMap = new Map<string, string[]>();

  for (const md of iterMd(docs)) {
    const content = fs.readFileSync(md, "utf-8");
    const [fm] = parseFrontmatter(content);
    const docId = String(fm["doc-id"] ?? "");
    if (!docId) continue;
    const rel = relTo(md, docs);
    if (!idMap.has(docId)) idMap.set(docId, []);
    idMap.get(docId)!.push(rel);
  }

  let dups = 0;
  for (const [did, paths] of idMap) {
    if (paths.length > 1) {
      warn("doc_id_dup", { doc_id: did, files: paths, ok: false });
      dups++;
    }
  }

  verbose("doc_id_done", {
    total: idMap.size,
    duplicates: dups,
    ok: dups === 0,
  });
}

// ── 8. API 파일 method/path 형식 ─────────────────────────────────────────────

export function checkApiFormat(docs: string): void {
  const apiDir = path.join(docs, "api");
  if (!fs.existsSync(apiDir)) {
    warn("api_format", { msg: "api/ 없음", ok: false });
    return;
  }

  const validMethods = new Set(["GET", "POST", "PUT", "PATCH", "DELETE"]);
  let badMethod = 0;
  let badPath = 0;

  for (const md of iterMd(apiDir)) {
    const content = fs.readFileSync(md, "utf-8");
    const [fm] = parseFrontmatter(content);
    if (!fm || fm["type"] !== "api-endpoint") continue;
    const rel = relTo(md, docs);
    const method = String(fm["method"] ?? "");
    const apiPath = String(fm["path"] ?? "");
    if (!validMethods.has(method)) {
      warn("api_method", { file: rel, method, ok: false });
      badMethod++;
    }
    if (!apiPath.startsWith("/")) {
      warn("api_path", { file: rel, path: apiPath, ok: false });
      badPath++;
    }
  }

  verbose("api_format_done", {
    bad_method: badMethod,
    bad_path: badPath,
    ok: badMethod === 0 && badPath === 0,
  });
}

File: scripts/review/checks.py
"""docs/ 전체 품질 검토 체크 함수 모음.

8개 체크 함수가 JSONL emit으로 결과를 출력한다.
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Iterator

try:
    import yaml
    _HAVE_YAML = True
except ImportError:
    _HAVE_YAML = False

from scripts.core.emit import info, warn, error
from scripts.core.paths import DST_DOCS

REQUIRED_FM_ALL = {"doc-id", "title", "description", "type", "status", "updated", "source"}
REQUIRED_FM_API = {"method", "path", "base-url", "category", "tags"}


# ── 파싱 헬퍼 ────────────────────────────────────────────────────────────────

def parse_frontmatter(content: str) -> tuple[dict, str]:
    """YAML frontmatter 파싱. (fm_dict, body) 반환."""
    if not content.startswith("---"):
        return {}, content
    end = content.find("\n---", 3)
    if end == -1:
        return {}, content
    fm_raw = content[3:end].strip()
    body = content[end + 4:].strip()
    if _HAVE_YAML:
        try:
            fm = yaml.safe_load(fm_raw) or {}
        except Exception:
            fm = {}
    else:
        # 단순 라인 파싱 폴백
        fm = {}
        for line in fm_raw.splitlines():
            m = re.match(r"^([a-zA-Z_-]+):\s*(.*)", line)
            if m:
                fm[m.group(1)] = m.group(2).strip().strip('"')
    return fm, body


def _iter_md(docs: Path) -> Iterator[Path]:
    """docs/ 하위 모든 .md 파일을 순회한다."""
    yield from sorted(docs.rglob("*.md"))


# ── 1. 파일 수 / 구조 ────────────────────────────────────────────────────────

def check_structure(docs: Path) -> None:
    dirs = {"api": docs / "api", "schema": docs / "schema",
            "category": docs / "category", "guide": docs / "guide"}
    total = 0
    counts: dict[str, int] = {}
    for name, d in dirs.items():
        files = sorted(d.rglob("*.md")) if d.exists() else []
        counts[name] = len(files)
        total += len(files)

    info("structure", total=total, **counts)

    for fname in ("llms.txt", "llms-full.txt"):
        p = docs / fname
        if p.exists():
            info("llms_file", file=fname, size=p.stat().st_size)
        else:
            warn("llms_file", file=fname, msg="파일 없음", ok=False)


# ── 2. 내부 링크 유효성 ───────────────────────────────────────────────────────

def check_links(docs: Path) -> int:
    """내부 링크(dead link) 검사. 발견된 dead link 수 반환."""
    dead = 0
    checked = 0
    link_pat = re.compile(r"\[([^\]]*)\]\(([^)#\s][^)\s]*)\)")

    for md in _iter_md(docs):
        content = md.read_text(encoding="utf-8")
        _, body = parse_frontmatter(content)
        for m in link_pat.finditer(body):
            href = m.group(2)
            if href.startswith(("http", "mailto", "#")):
                continue
            target = (md.parent / href).resolve()
            checked += 1
            if not target.exists():
                try:
                    rel = str(md.relative_to(docs))
                except ValueError:
                    rel = str(md)
                warn("dead_link", file=rel, href=href, ok=False)
                dead += 1

    info("links_done", checked=checked, dead=dead, ok=(dead == 0))
    return dead


# ── 3. frontmatter 완전성 ─────────────────────────────────────────────────────

def check_frontmatter(docs: Path) -> None:
    missing_required = 0
    missing_api = 0

    for md in _iter_md(docs):
        content = md.read_text(encoding="utf-8")
        fm, _ = parse_frontmatter(content)
        try:
            rel = str(md.relative_to(docs))
        except ValueError:
            rel = str(md)

        if not fm:
            warn("frontmatter", file=rel, msg="frontmatter 없음", ok=False)
            missing_required += 1
            continue

        page_type = fm.get("type", "")
        # guide 타입에서는 source 필드 없어도 허용
        required = REQUIRED_FM_ALL - ({"source"} if page_type == "guide" else set())
        missing = required - set(fm.keys())
        if missing:
            warn("frontmatter", file=rel,
                 msg=f"필수 필드 누락: {sorted(missing)}", ok=False)
            missing_required += 1

        if page_type == "api-endpoint":
            m2 = REQUIRED_FM_API - set(fm.keys())
            if m2:
                warn("frontmatter_api", file=rel,
                     msg=f"API 전용 필드 누락: {sorted(m2)}", ok=False)
                missing_api += 1

    info("frontmatter_done",
         missing_required=missing_required,
         missing_api=missing_api,
         ok=(missing_required == 0 and missing_api == 0))


# ── 4. 내용 이상 패턴 ────────────────────────────────────────────────────────

def check_content(docs: Path) -> None:
    html_pat = re.compile(
        r"<(?:div|span|table|tr|td|th|ul|li|p|br)\b", re.IGNORECASE
    )
    too_short = 0
    no_h1 = 0
    multi_h1 = 0
    has_html = 0

    for md in _iter_md(docs):
        content = md.read_text(encoding="utf-8")
        _, body = parse_frontmatter(content)
        try:
            rel = str(md.relative_to(docs))
        except ValueError:
            rel = str(md)

        if len(body.strip()) < 100:
            warn("content", file=rel, msg="본문 100자 미만", ok=False)
            too_short += 1

        h1s = re.findall(r"^# .+", body, re.MULTILINE)
        if len(h1s) == 0:
            warn("content", file=rel, msg="H1 없음", ok=False)
            no_h1 += 1
        elif len(h1s) > 1:
            warn("content", file=rel, msg=f"H1 중복: {len(h1s)}개", ok=False)
            multi_h1 += 1

        if html_pat.search(body):
            warn("content", file=rel, msg="HTML 태그 잔존", ok=False)
            has_html += 1

    info("content_done",
         too_short=too_short,
         no_h1=no_h1,
         multi_h1=multi_h1,
         has_html=has_html,
         ok=(too_short == 0 and no_h1 == 0 and multi_h1 == 0 and has_html == 0))


# ── 5. llms.txt 정합성 ───────────────────────────────────────────────────────

def check_llms(docs: Path) -> None:
    llms = docs / "llms.txt"
    if not llms.exists():
        error("llms", file="llms.txt", msg="파일 없음", ok=False)
        return

    content = llms.read_text(encoding="utf-8")
    links = re.findall(r"\(([^)]+\.md)\)", content)
    broken = []
    for link in links:
        target = (docs / link).resolve()
        if not target.exists():
            broken.append(link)

    sections = re.findall(r"^## .+", content, re.MULTILINE)
    entries = re.findall(r"^- \[", content, re.MULTILINE)

    if broken:
        for b in broken:
            warn("llms_link", file="llms.txt", href=b, ok=False)

    info("llms_done",
         sections=len(sections),
         entries=len(entries),
         links=len(links),
         broken=len(broken),
         ok=(len(broken) == 0))


# ── 6. 카테고리 인덱스 링크 ──────────────────────────────────────────────────

def check_category_links(docs: Path) -> None:
    cat_dir = docs / "category"
    if not cat_dir.exists():
        warn("category_links", msg="category/ 없음", ok=False)
        return

    total = 0
    dead = 0
    link_pat = re.compile(r"\[([^\]]+)\]\(([^)]+)\)")

    for md in sorted(cat_dir.glob("*.md")):
        content = md.read_text(encoding="utf-8")
        _, body = parse_frontmatter(content)
        try:
            rel = str(md.relative_to(docs))
        except ValueError:
            rel = str(md)
        for m in link_pat.finditer(body):
            href = m.group(2)
            if href.startswith(("http", "#")):
                continue
            total += 1
            target = (md.parent / href).resolve()
            if not target.exists():
                warn("category_link", file=rel, href=href, ok=False)
                dead += 1

    info("category_links_done", total=total, dead=dead, ok=(dead == 0))


# ── 7. doc-id 중복 ───────────────────────────────────────────────────────────

def check_doc_id_uniqueness(docs: Path) -> None:
    id_map: dict[str, list[str]] = {}
    for md in _iter_md(docs):
        content = md.read_text(encoding="utf-8")
        fm, _ = parse_frontmatter(content)
        doc_id = str(fm.get("doc-id", "")) if fm else ""
        if doc_id:
            try:
                rel = str(md.relative_to(docs))
            except ValueError:
                rel = str(md)
            id_map.setdefault(doc_id, []).append(rel)

    dups = {k: v for k, v in id_map.items() if len(v) > 1}
    for did, paths in dups.items():
        warn("doc_id_dup", doc_id=did, files=paths, ok=False)

    info("doc_id_done",
         total=len(id_map),
         duplicates=len(dups),
         ok=(len(dups) == 0))


# ── 8. API 파일 method/path 형식 ─────────────────────────────────────────────

def check_api_format(docs: Path) -> None:
    api_dir = docs / "api"
    if not api_dir.exists():
        warn("api_format", msg="api/ 없음", ok=False)
        return

    valid_methods = {"GET", "POST", "PUT", "PATCH", "DELETE"}
    bad_method = 0
    bad_path = 0

    for md in sorted(api_dir.rglob("*.md")):
        content = md.read_text(encoding="utf-8")
        fm, _ = parse_frontmatter(content)
        if not fm or fm.get("type") != "api-endpoint":
            continue
        try:
            rel = str(md.relative_to(docs))
        except ValueError:
            rel = str(md)
        method = str(fm.get("method", ""))
        path = str(fm.get("path", ""))
        if method not in valid_methods:
            warn("api_method", file=rel, method=method, ok=False)
            bad_method += 1
        if not path.startswith("/"):
            warn("api_path", file=rel, path=path, ok=False)
            bad_path += 1

    info("api_format_done",
         bad_method=bad_method,
         bad_path=bad_path,
         ok=(bad_method == 0 and bad_path == 0))
